web/src/page/admin/index.tsx

```diff
@@ -1,104 +1,309 @@
+  const [modalState, setModalState] = React.useState<String>("");
+  const [confirmLoading, setConfirmLoading] = React.useState(false);
+  const [modalData, setModalData] = React.useState<IUser>(INITIAL_USER_STATE);
+  const [userData, setUserData] = React.useState<IUserList>([]);
+
+  useEffect(() => {
+    Service.get("/v1/admin/user/").then((response) => setUserData(response));
+  }, []);
 
+  const showModal = (record: IUser) => {
+    setModalState(MODAL_STATES.EDIT);
+    setModalData(record);
+  };
+  const [form] = Form.useForm();
 
-const columns = [
-  {
-    title: 'Name',
-    dataIndex: 'name',
-    key: 'name',
-    render: (text: any) => <span>{text}</span>,
-  },
-  {
-    title: 'Email',
-    dataIndex: 'email',
-    key: 'email',
-  },
-  {
-    title: 'Status',
-    dataIndex: 'status',
-    key: 'status',
-    render: (status: any) => status === "Active" ? <Switch defaultChecked disabled /> : <Switch disabled />,
-  },
-  {
-    title: 'Roles',
-    key: 'roles',
-    dataIndex: 'roles',
-    render: (tags: any) => (
-      <>
-        {tags.map((tag: any) => {
-          let color = tag.length > 5 ? 'geekblue' : 'green';
-          if (tag === 'admin') {
-            color = 'volcano';
-          }
-          return (
-            <Tag color={color} key={tag}>
-              {tag.toUpperCase()}
-            </Tag>
+  function onCreateuserClick() {
+    form.resetFields();
+    setModalState(MODAL_STATES.CREATE);
+  }
+
+  const onFormSubmit = (values: IUser) => {
+    function updateData(response: IUser) {
+      setUserData((state: IUserList) => {
+        if (modalState === MODAL_STATES.CREATE) {
+          return [...state, response];
+        } else {
+          const dataIndex = state.findIndex(
+            (user: IUser) => user.id === values.id
           );
-        })}
-      </>
-    ),
-  },
-  {
-    title: 'Action',
-    key: 'action',
-    render: (text: string, record: any) =>
-    (
-      <Space size="middle">
-        <span>Edit</span>
-        <Popconfirm
-          title="Are you sure want to delete this user?"
-          onConfirm={confirm}
-          // onCancel={cancel}
-          okText="Delete"
-          cancelText="Cancel"
-        >
-          <a href="/" className={styles.userDelete}>Delete</a>
-        </Popconfirm>
+          const cloneData = { ...values };
+          state[dataIndex] = cloneData;
+          state[dataIndex].roles = [cloneData.roles].flat();
+          return [...state];
+        }
+      });
+    }
+    let url: String = "/v1/admin/user/";
+    let promise = Service.post;
+    if (modalState !== MODAL_STATES.CREATE) {
+      url = `/v1/admin/user/${values.id}`;
+      promise = Service.update;
+    }
+    const body = { ...values };
+    promise(url, { body })
+      .then((response: IUser) => {
+        updateData(response);
+      })
+      .catch(() => {
+        updateData(INITIAL_USER_STATE);
+      });
+    setModalData(INITIAL_USER_STATE);
```

Comment:
move this to a separate function called resetStates

Reply:
updated